File: solutions/javascript/anagram/1/anagram.js
export const findAnagrams = (subject, potentialAnagrams) => {
  return potentialAnagrams.filter((candidate) => isAnagram(candidate, subject));
};

const isAnagram = (word1, word2) => {
  const lowerCaseWord1 = word1.toLowerCase()
  const lowercaseWord2 = word2.toLowerCase()
  return (
    lowerCaseWord1 !== lowercaseWord2 &&
    sortedLetters(lowerCaseWord1) === sortedLetters(lowercaseWord2)
  );
};

const sortedLetters = (word) => {
  return [...word].sort().join();
};

File: solutions/r/anagram/1/anagram.R
anagram <- function(subject, candidates) {
  filter <- vector("logical", length(candidates))
  for (i in seq_along(candidates)) {
    filter[i] <- is_anagram(subject, candidates[i])
  }
  if (length(candidates[filter]) == 0) {
    NULL
  } else {
    candidates[filter]
  }
}

is_anagram <- function(word1, word2) {
  tolower(word1) != tolower(word2) & sorted_letters(word1) == sorted_letters(word2)
}

sorted_letters <- function(word) {
  word |>
  tolower() |>
  strsplit("") |>
  unlist() |>
  sort() |>
  paste(collapse="")
}
